k4FWCore/components/EfficiencyFilter.cpp

```diff
@@ -0,0 +1,137 @@
+    StatusCode sc = Transformer::initialize();
+    if (sc.isFailure()) {
+      error() << "Unable to initialize base class Service." << endmsg;
+      return sc;
+    }
+
+    m_uniqueIDSvc = service("UniqueIDGenSvc");
+    if (!m_uniqueIDSvc) {
+      error() << "Unable to locate the UniqueIDGenSvc" << endmsg;
+      return StatusCode::FAILURE;
+    }
+
+    if (m_efficiency < 0.0 || m_efficiency > 1.0) {
+      error() << "Efficiency must be between 0 and 1." << endmsg;
+      return StatusCode::FAILURE;
+    }
+
+    return StatusCode::SUCCESS;
+  }
+
+  podio::CollectionBase* operator()(const podio::CollectionBase& coll,
+                                    const edm4hep::EventHeaderCollection& evtHeader) const override {
+
+    debug() << "EfficiencyFilter called for collection of type " << coll.getTypeName() << endmsg;
+
+    const auto uid = m_uniqueIDSvc->getUniqueID(evtHeader, name());
+
+    auto ptr = dispatchByType(
+        coll, uid,
+        edm4hep::edm4hepDataTypes{}); // Dispatch to the correct createSubsetCollection template function based on type
+    if (!ptr) {
+      ptr = dispatchByType(coll, uid, edm4hep::edm4hepLinkTypes{}); // Try link types if not found in data types
+    }
+    if (!ptr) {
+      throw std::runtime_error(fmt::format(
+          "EfficiencyFilter: No createSubsetCollection function found for collection type '{}'", coll.getTypeName()));
+    }
+    return ptr;
+  }
+
+private:
+  SmartIF<IUniqueIDGenSvc> m_uniqueIDSvc{nullptr};
+
+  Gaudi::Property<double> m_efficiency{this, "Efficiency", 0.5,
+                                       "The efficiency of the filter, between 0 and 1. If 0.5, each item has a 50% "
+                                       "chance to be kept in the output collection."};
+  Gaudi::Property<bool> m_exact{
+      this, "Exact", true,
+      "If true, the output collection will have exactly the number of elements given by the efficiency. For example, "
+      "if the efficiency is 0.5 and the input collection has 10 elements, the output collection will have exactly 5 "
+      "elements. The result is floored."
+      "If false, the output collection will have a random number of elements, with the average number of elements "
+      "given by the efficiency."};
+
+  template <typename... T>
+  podio::CollectionBase* dispatchByType(const podio::CollectionBase& coll, size_t uid,
+                                        podio::utils::TypeList<T...>&&) const {
+    std::string_view typeName = coll.getTypeName();
+    podio::CollectionBase* result = nullptr;
+    (..., (typeName == T::collection_type::typeName
+               ? (result = createSubsetCollection<typename T::collection_type>(&coll, uid))
+               : nullptr));
+    return result;
+  }
+
+  template <typename T>
+  podio::CollectionBase* createSubsetCollection(const podio::CollectionBase* source, const size_t uid) const {
+
+    std::mt19937 randomGen(std::random_device{}());
+    randomGen.seed(uid);
```

Comment:
```suggestion
    std::mt19937 randomGen(uid);
```

Seed can be set directly in the constructor